### PR TITLE
npm - handle json decode exception

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

* Provide a user friendly message by handling json decode
  exception rather than providing a stacktrace

Fixes: #1614

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/1614_npm.yml
plugins/modules/packaging/language/npm.py

### CODE REVIEW

File: plugins/modules/packaging/language/npm.py [hunk 1]
```diff
@@ -55,12 +55,12 @@
     default: no
   ci:
     description:
-      - Install packages based on package-lock file, same as running npm ci
+      - Install packages based on package-lock file, same as running npm ci.
```
Comment (felixfontein) [inline plugins/modules/packaging/language/npm.py:58]:
```suggestion
      - Install packages based on package-lock file, same as running C(npm ci).
```

File: plugins/modules/packaging/language/npm.py [hunk 2]
```diff
@@ -193,14 +193,18 @@ def _exec(self, args, run_in_check_mode=False, check_rc=True):
 
             rc, out, err = self.module.run_command(cmd, check_rc=check_rc, cwd=cwd)
             return out
-        return ''
+        return '{}'
```
Comment (felixfontein) [inline plugins/modules/packaging/language/npm.py:196]: Is this change necessary?

File: plugins/modules/packaging/language/npm.py [hunk 3]
```diff
@@ -174,7 +174,7 @@ def _exec(self, args, run_in_check_mode=False, check_rc=True):
                 cmd.append('--ignore-scripts')
             if self.unsafe_perm:
                 cmd.append('--unsafe-perm')
-            if self.name:
+            if self.name and 'list' not in cmd:
```
Comment (felixfontein) [inline plugins/modules/packaging/language/npm.py:177]: Why not have an additional optional parameter `add_package_name=True` which is set to `False` in the invocation inside `list()`?

File: plugins/modules/packaging/language/npm.py (hunk 2 above)
Comment (Akasurde) [reply plugins/modules/packaging/language/npm.py:196]: yes

```
>>> a = ''
>>> import json
>>> json.loads(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
>>> json.loads('{}')
{}
```

File: plugins/modules/packaging/language/npm.py (hunk 3 above)
Comment (Akasurde) [reply plugins/modules/packaging/language/npm.py:177]: ok.

File: plugins/modules/packaging/language/npm.py (hunk 2 above)
Comment (felixfontein) [reply plugins/modules/packaging/language/npm.py:196]: There's only one piece of code that uses `json.loads` of the result, and that code passes `run_in_check_mode=True`. When `run_in_check_mode` is `True`, this line is never hit.

Comment (felixfontein) [reply plugins/modules/packaging/language/npm.py:196]: (Also a way better workaround would be `data = json.loads(self._exec(cmd, True, False) or '{}')` instead of modifying the general return value of `self._exec`.)

**Review by felixfontein — APPROVED** [review]

File: plugins/modules/packaging/language/npm.py [hunk 4]
```diff
@@ -174,7 +174,7 @@ def _exec(self, args, run_in_check_mode=False, check_rc=True):
                 cmd.append('--ignore-scripts')
             if self.unsafe_perm:
                 cmd.append('--unsafe-perm')
-            if self.name:
+            if self.name and add_package_name:
```
Comment (shaharmor) [inline plugins/modules/packaging/language/npm.py:177]: This change breaks all version specific updates because this was the way npm was checking if a specific version is installed.

Before:
`npm list --json --long pkg@1.2.3` would return `{}` if `pkg` was not installed with version `1.2.3`

After
`npm list --json --long` would return the full `pkg` info regardless of which version is installed, and this will make ansible think that the module is installed.

Comment (felixfontein) [reply plugins/modules/packaging/language/npm.py:177]: Yes, it looks like you are correct. (Though I wouldn't say that it is this change, but the call of this function with `add_package_name=False`.) So either it passes `add_package_name=False`, or it has special code in `list()` to check for the correct version.

This is also relevant for `state=absent`, since before the module would only uninstall the package if the version matches (assuming the version is provided), while now it will always uninstall it and ignore the version. The new behavior better matches the documentation IMO, but this is still different from before.

@Akasurde can you take a look?

Comment (felixfontein) [reply plugins/modules/packaging/language/npm.py:177]: @Akasurde ping ^
